### DISCUSSION TITLE
Upgrade go  from 1.16 ?

### DISCUSSION THREAD

**Discussion by Johnlon** [cucumber/godog#660] [Ideas]
@vearutop 
@mpkorstanje 

Any reason why we need to stick on 1.16?

I want to use generics internally and need 1.18 and a bunch of basic utilities are missing in our ancient 1.16.

Summary ... 
1.16 Feb 2021 - is ancient history
1.19 Aug 2022 - upgrade to this at least, utilities libs, generics
1.20 Feb 2023 - adds potentially useful conversions
1.21 Aug 2023 - adds potentially useful type inference
1.22 Feb 2024 - improved for loops, linter improvements
1.23 Aug 2024 - improved iterators

**Comment by vearutop** [comment]
That's a good point, I've moved required status checks to [stable/oldstable](https://github.com/actions/setup-go?tab=readme-ov-file#using-stableoldstable-aliases) Go versions. I think we can remove 1.16 and 1.17 from CI.

**Reply by Johnlon** [reply]
Does that mean I can change go.mod to 1.22?

**Reply by Johnlon** [reply]
or 1.23?

**Reply by vearutop** [reply]
Let's bump it to what's needed? E.g. if you want generics in the `godog` itself (not talking about the users of the library), that'd be 1.18.

Bumping too high can be an unpleasant upgrade demand to those who lag behind. I think it is generally a good idea to keep backwards compatibility when that's feasible and doesn't incur high maintenance costs.

**Comment by Johnlon** [comment]
some basic API missing too

on syrings and sort and slice

**Comment by Johnlon** [comment]
2 years out of date is probably a cut off I think.  (IE 1.19)

I'll see what cuts the mustard.

-++

btw the old honnef we use 0.4.6  crashes on my 1.22 install (or is it 1.23?) so upgrading to 0.5.1 fixed that build issue but won't work on 1.16

**Comment by kcross-ctoken** [comment]
hmm si this where `../../go/pkg/mod/golang.org/x/tools@v0.21.1-0.20240531212143-b6235391adb3/internal/tokeninternal/tokeninternal.go:64:9: invalid array length -delta * delta (constant -256 of type int64)` comes from ?

**Comment by kcross-ctoken** [comment]
BTW ... if we update the gherkin spec we should make it a major relase as its a breaking change in a lot of cases.
I know our tests now have to be update to accomodate the new gherkin spec.

**Comment by kcross-ctoken** [comment]
Id like to introduce something like mise to enable a consistent environment when running the tests ect.
https://github.com/jdx/mise
I struggled to figure out what version of go works with the tests while doing my recent pr.
If its all in a config file then its pretty easy to work it out .
mise install ... and you have all the relevant versions to run.
Id also like to use golantci-lint that is relevant to the version of go so we can lint out the code properly
